### DISCUSSION TITLE
Multiple mTLS Client Auth on Routes

### DISCUSSION THREAD

**Discussion by msmost** [apache/apisix#8799] [Q&A]
After reading the [Configure mTLS for client to APISIX](https://apisix.apache.org/docs/apisix/tutorials/client-to-apisix-mtls/) tutorial, the options for configuring mTLS fall short of our desired use case.

Our use case would have APISIX:
* Support client certificate authentication _on a specific `route`_.

  The tutorial shows that during the [Configure the `route` in APISIX](https://apisix.apache.org/docs/apisix/tutorials/client-to-apisix-mtls/#configure-the-route-in-apisix) step, there isn't any part of the `route` configuration that ties back to the mTLS setup. We conclude that when mTLS is configured ALL `route`s are affected.

  We searched for a plugin that would assist in this configuration and weren't able to find one.

* Support _multiple_ client certificates to support authentication for multiple clients.
  The tutorial shows how to add a single client certificate to an SSL object ([Configure the certificate in APISIX](https://apisix.apache.org/docs/apisix/tutorials/client-to-apisix-mtls/#configure-the-certificate-in-apisix)). However, we need to be able to authenticate multiple clients using different certificates.

**Comment by tokers** [comment]
@msmost Currently, the route-level mTLS support is lack for APISIX.

> The tutorial shows how to add a single client certificate to an SSL object ([Configure the certificate in APISIX](https://apisix.apache.org/docs/apisix/tutorials/client-to-apisix-mtls/#configure-the-certificate-in-apisix)). However, we need to be able to authenticate multiple clients using different certificates.

Did you mean, for different clients, you may use different CA certificate to verify their cert?

**Reply by msmost** [reply]
@tokers 
> Did you mean, for different clients, you may use different CA certificate to verify their cert?

Correct. Each client would have a cert signed by different CA certificates. An example would be when we do not have possession of the CA certificate. Self-signed certificates can be used for setting up mTLS as long as the public key has been shared with us for verification.

**Reply by tokers** [reply]
Got it. So the concern will be how to choose the correct CA when doing the TLS handshake.

**Reply by msmost** [reply]
Has there been any other desire from the community to establish route level configuration support for mTLS?

**Reply by yelga** [reply]
+1 Would love to have route-level mTLS support, we have the exact same use-case.

**Reply by deekshith-hadil** [reply]
+1 We have the same use-case. Looking for this feature.

**Reply by juzhiyuan** [reply]
🤔 before making this available in apisix, one possible workaround is to use different SNIs:

- sni-mtls.example.com for mTLS cases
- sni-common.example.com for other cases.

I'm unsure if this should be supported by APISIX natively... (cc @bzp2010 @moonming )

**Reply by deekshith-hadil** [reply]
Yes, SNI can serve as a workaround, but our primary requirement is a TCP proxy server suitable for use in a medical device environment. In this ecosystem, communications are conducted predominantly using IP addresses. Multiple client devices connect through APISIX as a proxy, and it is essential that APISIX is able to verify each individual client during these interactions.
